// vue/src/core/observer/dep.js
/* @flow */

import type Watcher from './watcher'
import { remove } from '../util/index'

let uid = 0

/**
 * A dep is an observable that can have multiple
 * directives subscribing to it.
 */
// 1-2-5-2-3-2-3-2-5-1 dep 相当于媒介 将观察者 和 目标数据关联起来
export default class Dep {
  static target: ?Watcher;
  id: number;
  subs: Array<Watcher>;

  constructor () {
    this.id = uid++
    this.subs = []
  }
  // 1-2-5-2-3-2-3-2-5-2 将 观察者(订阅者)们 收集起来
  addSub (sub: Watcher) {
    this.subs.push(sub)
  }

  removeSub (sub: Watcher) {
    remove(this.subs, sub)
  }
  // 1-2-5-2-3-2-3-2-5-3 给 Dep 的 观察者们 关联 媒介实例
  depend () {
    if (Dep.target) {
      Dep.target.addDep(this)
    }
  }
  // 1-2-5-2-3-2-3-2-5-4 更新 target watcher being evaluated
  notify () {
    // stabilize the subscriber list first
    const subs = this.subs.slice()
    for (let i = 0, l = subs.length; i < l; i++) {
      // 1-2-5-2-3-2-3-2-5-5 在 src/core/observer/watcher.js 中定义
      subs[i].update()
    }
  }
}

// the current target watcher being evaluated.
// this is globally unique because there could be only one
// watcher being evaluated at any time.
Dep.target = null
const targetStack = []

export function pushTarget (_target: Watcher) {
  if (Dep.target) targetStack.push(Dep.target)
  Dep.target = _target
}

export function popTarget () {
  Dep.target = targetStack.pop()
}
