// vux/src/components/index.js
const vux = {}
/* only for building vux.css */
import Style from '../styles/index.vue'
import vuxActionsheet from 'src/components/actionsheet/index.vue'

vux['vuxActionsheet'] = vuxActionsheet
import vuxAjaxPlugin from 'src/plugins/ajax/index.js'

vux['vuxAjaxPlugin'] = vuxAjaxPlugin
import vuxAlert from 'src/components/alert/index.vue'

vux['vuxAlert'] = vuxAlert
import vuxAlertPlugin from 'src/plugins/alert/index.js'

vux['vuxAlertPlugin'] = vuxAlertPlugin
import vuxArray2stringFilter from 'src/filters/array2String.js'

vux['vuxArray2stringFilter'] = vuxArray2stringFilter
import vuxBadge from 'src/components/badge/index.vue'

vux['vuxBadge'] = vuxBadge
import vuxbase64 from 'src/tools/base64/index.js'

vux['vuxbase64'] = vuxbase64
import vuxBlur from 'src/components/blur/index.vue'

vux['vuxBlur'] = vuxBlur
import vuxBox from 'src/components/box/index.vue'

vux['vuxBox'] = vuxBox
import vuxButtonTab from 'src/components/button-tab/button-tab.vue'

vux['vuxButtonTab'] = vuxButtonTab
import vuxButtonTabItem from 'src/components/button-tab/button-tab-item.vue'

vux['vuxButtonTabItem'] = vuxButtonTabItem
import vuxCalendar from 'src/components/calendar/index.vue'

vux['vuxCalendar'] = vuxCalendar
import vuxCard from 'src/components/card/index.vue'

vux['vuxCard'] = vuxCard
import vuxCell from 'src/components/cell/index.vue'

vux['vuxCell'] = vuxCell
import vuxChecker from 'src/components/checker/checker.vue'

vux['vuxChecker'] = vuxChecker
import vuxCheckerItem from 'src/components/checker/checker-item.vue'

vux['vuxCheckerItem'] = vuxCheckerItem
import vuxCheckIcon from 'src/components/check-icon/index.vue'

vux['vuxCheckIcon'] = vuxCheckIcon
import vuxChecklist from 'src/components/checklist/index.vue'

vux['vuxChecklist'] = vuxChecklist
import vuxChinaAddressData from 'src/datas/china_address.json'

vux['vuxChinaAddressData'] = vuxChinaAddressData
import vuxClickOutsideDirective from 'src/directives/click-outside/index.js'

vux['vuxClickOutsideDirective'] = vuxClickOutsideDirective
import vuxClocker from 'src/components/clocker/index.vue'

vux['vuxClocker'] = vuxClocker
import vuxColorPicker from 'src/components/color-picker/index.vue'

vux['vuxColorPicker'] = vuxColorPicker
import vuxConfirm from 'src/components/confirm/index.vue'

vux['vuxConfirm'] = vuxConfirm
import vuxConfirmPlugin from 'src/plugins/confirm/index.js'

vux['vuxConfirmPlugin'] = vuxConfirmPlugin
import vuxcookie from 'src/tools/cookie/index.js'

vux['vuxcookie'] = vuxcookie
import vuxCountdown from 'src/components/countdown/index.vue'

vux['vuxCountdown'] = vuxCountdown
import vuxCountup from 'src/components/countup/index.vue'

vux['vuxCountup'] = vuxCountup
import vuxdateFormat from 'src/tools/date/format.js'

vux['vuxdateFormat'] = vuxdateFormat
import vuxDatetime from 'src/components/datetime/index.vue'

vux['vuxDatetime'] = vuxDatetime
import vuxdebounce from 'src/tools/debounce/index.js'

vux['vuxdebounce'] = vuxdebounce
import vuxDemobasic from 'src/components/fullpage/DemoBasic.vue'

vux['vuxDemobasic'] = vuxDemobasic
import vuxDemoindex from 'src/components/popover/DemoIndex.vue'

vux['vuxDemoindex'] = vuxDemoindex
import vuxDevicePlugin from 'src/plugins/device/index.js'

vux['vuxDevicePlugin'] = vuxDevicePlugin
import vuxDevTip from 'src/components/dev-tip/index.vue'

vux['vuxDevTip'] = vuxDevTip
import vuxDivider from 'src/components/divider/index.vue'

vux['vuxDivider'] = vuxDivider
import vuxFlexbox from 'src/components/flexbox/flexbox.vue'

vux['vuxFlexbox'] = vuxFlexbox
import vuxFlexboxItem from 'src/components/flexbox/flexbox-item.vue'

vux['vuxFlexboxItem'] = vuxFlexboxItem
import vuxFormatTimeFilter from 'src/filters/format-time.js'

vux['vuxFormatTimeFilter'] = vuxFormatTimeFilter
import vuxFormPreview from 'src/components/form-preview/index.vue'

vux['vuxFormPreview'] = vuxFormPreview
import vuxFriendlyTimeFilter from 'src/filters/friendly-time.js'

vux['vuxFriendlyTimeFilter'] = vuxFriendlyTimeFilter
import vuxFullpage from 'src/components/fullpage/index.vue'

vux['vuxFullpage'] = vuxFullpage
import vuxGrid from 'src/components/grid/grid.vue'

vux['vuxGrid'] = vuxGrid
import vuxGridItem from 'src/components/grid/grid-item.vue'

vux['vuxGridItem'] = vuxGridItem
import vuxGroup from 'src/components/group/index.vue'

vux['vuxGroup'] = vuxGroup
import vuxGroupTitle from 'src/components/group-title/index.vue'

vux['vuxGroupTitle'] = vuxGroupTitle
import vuxIcon from 'src/components/icon/index.vue'

vux['vuxIcon'] = vuxIcon
import vuxInlineCalendar from 'src/components/inline-calendar/index.vue'

vux['vuxInlineCalendar'] = vuxInlineCalendar
import vuxInlineDesc from 'src/components/inline-desc/index.vue'

vux['vuxInlineDesc'] = vuxInlineDesc
import vuxInlineXNumber from 'src/components/inline-x-number/index.vue'

vux['vuxInlineXNumber'] = vuxInlineXNumber
import vuxInviewDirective from 'src/directives/inview/index.js'

vux['vuxInviewDirective'] = vuxInviewDirective
import vuxLoading from 'src/components/loading/index.vue'

vux['vuxLoading'] = vuxLoading
import vuxLoadingPlugin from 'src/plugins/loading/index.js'

vux['vuxLoadingPlugin'] = vuxLoadingPlugin
import vuxLoadMore from 'src/components/load-more/index.vue'

vux['vuxLoadMore'] = vuxLoadMore
import vuxMasker from 'src/components/masker/index.vue'

vux['vuxMasker'] = vuxMasker
import vuxmd5 from 'src/tools/md5/index.js'

vux['vuxmd5'] = vuxmd5
import vuxName2valueFilter from 'src/filters/name2value.js'

vux['vuxName2valueFilter'] = vuxName2valueFilter
import vuxnumberComma from 'src/tools/number/comma.js'

vux['vuxnumberComma'] = vuxnumberComma
import vuxnumberPad from 'src/tools/number/pad.js'

vux['vuxnumberPad'] = vuxnumberPad
import vuxnumberRandom from 'src/tools/number/random.js'

vux['vuxnumberRandom'] = vuxnumberRandom
import vuxNumberRoller from 'src/components/number-roller/index.vue'

vux['vuxNumberRoller'] = vuxNumberRoller
import vuxPanel from 'src/components/panel/index.vue'

vux['vuxPanel'] = vuxPanel
import vuxPicker from 'src/components/picker/index.vue'

vux['vuxPicker'] = vuxPicker
import vuxPopover from 'src/components/popover/index.vue'

vux['vuxPopover'] = vuxPopover
import vuxPopup from 'src/components/popup/index.vue'

vux['vuxPopup'] = vuxPopup
import vuxPopupPicker from 'src/components/popup-picker/index.vue'

vux['vuxPopupPicker'] = vuxPopupPicker
import vuxPreviewer from 'src/components/previewer/index.vue'

vux['vuxPreviewer'] = vuxPreviewer
import vuxQrcode from 'src/components/qrcode/index.vue'

vux['vuxQrcode'] = vuxQrcode
import vuxquerystring from 'src/tools/querystring/index.js'

vux['vuxquerystring'] = vuxquerystring
import vuxRadio from 'src/components/radio/index.vue'

vux['vuxRadio'] = vuxRadio
import vuxRange from 'src/components/range/index.vue'

vux['vuxRange'] = vuxRange
import vuxRater from 'src/components/rater/index.vue'

vux['vuxRater'] = vuxRater
import vuxScroller from 'src/components/scroller/index.vue'

vux['vuxScroller'] = vuxScroller
import vuxSearch from 'src/components/search/index.vue'

vux['vuxSearch'] = vuxSearch
import vuxSelector from 'src/components/selector/index.vue'

vux['vuxSelector'] = vuxSelector
import vuxShake from 'src/components/shake/index.vue'

vux['vuxShake'] = vuxShake
import vuxSpinner from 'src/components/spinner/index.vue'

vux['vuxSpinner'] = vuxSpinner
import vuxStep from 'src/components/step/step.vue'

vux['vuxStep'] = vuxStep
import vuxStepItem from 'src/components/step/step-item.vue'

vux['vuxStepItem'] = vuxStepItem
import vuxSticky from 'src/components/sticky/index.vue'

vux['vuxSticky'] = vuxSticky
import vuxstringTrim from 'src/tools/string/trim.js'

vux['vuxstringTrim'] = vuxstringTrim
import vuxSwipeout from 'src/components/swipeout/swipeout.vue'

vux['vuxSwipeout'] = vuxSwipeout
import vuxSwipeoutButton from 'src/components/swipeout/swipeout-button.vue'

vux['vuxSwipeoutButton'] = vuxSwipeoutButton
import vuxSwipeoutItem from 'src/components/swipeout/swipeout-item.vue'

vux['vuxSwipeoutItem'] = vuxSwipeoutItem
import vuxSwiper from 'src/components/swiper/swiper.vue'

vux['vuxSwiper'] = vuxSwiper
import vuxSwiperItem from 'src/components/swiper/swiper-item.vue'

vux['vuxSwiperItem'] = vuxSwiperItem
import vuxTab from 'src/components/tab/tab.vue'

vux['vuxTab'] = vuxTab
import vuxTabbar from 'src/components/tabbar/tabbar.vue'

vux['vuxTabbar'] = vuxTabbar
import vuxTabbarItem from 'src/components/tabbar/tabbar-item.vue'

vux['vuxTabbarItem'] = vuxTabbarItem
import vuxTabItem from 'src/components/tab/tab-item.vue'

vux['vuxTabItem'] = vuxTabItem
import vuxthrottle from 'src/tools/throttle/index.js'

vux['vuxthrottle'] = vuxthrottle
import vuxTimeline from 'src/components/timeline/timeline.vue'

vux['vuxTimeline'] = vuxTimeline
import vuxTimelineItem from 'src/components/timeline/timeline-item.vue'

vux['vuxTimelineItem'] = vuxTimelineItem
import vuxTip from 'src/components/tip/index.vue'

vux['vuxTip'] = vuxTip
import vuxTitlePlugin from 'src/plugins/title/index.js'

vux['vuxTitlePlugin'] = vuxTitlePlugin
import vuxToast from 'src/components/toast/index.vue'

vux['vuxToast'] = vuxToast
import vuxToastPlugin from 'src/plugins/toast/index.js'

vux['vuxToastPlugin'] = vuxToastPlugin
import vuxTransferDomDirective from 'src/directives/transfer-dom/index.js'

vux['vuxTransferDomDirective'] = vuxTransferDomDirective
import vuxValue2nameFilter from 'src/filters/value2name.js'

vux['vuxValue2nameFilter'] = vuxValue2nameFilter
import vuxVideo from 'src/components/video/index.vue'

vux['vuxVideo'] = vuxVideo
import vuxViewBox from 'src/components/view-box/index.vue'

vux['vuxViewBox'] = vuxViewBox
import vuxVuxComponentListData from 'src/datas/vux_component_list.json'

vux['vuxVuxComponentListData'] = vuxVuxComponentListData
import vuxWechatEmotion from 'src/components/wechat-emotion/index.vue'

vux['vuxWechatEmotion'] = vuxWechatEmotion
import vuxWechatPlugin from 'src/plugins/wechat/index.js'

vux['vuxWechatPlugin'] = vuxWechatPlugin
import vuxXAddress from 'src/components/x-address/index.vue'

vux['vuxXAddress'] = vuxXAddress
import vuxXButton from 'src/components/x-button/index.vue'

vux['vuxXButton'] = vuxXButton
import vuxXCircle from 'src/components/x-circle/index.vue'

vux['vuxXCircle'] = vuxXCircle
import vuxXDialog from 'src/components/x-dialog/index.vue'

vux['vuxXDialog'] = vuxXDialog
import vuxXHeader from 'src/components/x-header/index.vue'

vux['vuxXHeader'] = vuxXHeader
import vuxXImg from 'src/components/x-img/index.vue'

vux['vuxXImg'] = vuxXImg
import vuxXInput from 'src/components/x-input/index.vue'

vux['vuxXInput'] = vuxXInput
import vuxXNumber from 'src/components/x-number/index.vue'

vux['vuxXNumber'] = vuxXNumber
import vuxXProgress from 'src/components/x-progress/index.vue'

vux['vuxXProgress'] = vuxXProgress
import vuxXSwitch from 'src/components/x-switch/index.vue'

vux['vuxXSwitch'] = vuxXSwitch
import vuxXTextarea from 'src/components/x-textarea/index.vue'

vux['vuxXTextarea'] = vuxXTextarea

if (!!window) {
  for (let i in vux) {
    window[i] = vux[i]
  }
}
module.exports = vux
